api/nakadi-event-bus-api.yaml

```diff
@@ -1503,11 +1541,37 @@ definitions:
         type: string
         default: 'random'
 
+      compatibility_mode:
+        description: |
+          Compatibility modes provide a mean for event owners to evolve their schema, given changes
+          respect the semantics defined by this field.
+
+          It's designed to be flexible enough so that producers can evolve their schemas while not
+          inadvertently breaking existent consumers.
+
+          Once defined, the compatibility mode is fixed, since otherwise it would break a predefined contract,
+          declared by the producer.
+
+          List of available compatibility modes:
+
+          - 'deprecated': Schema changes are not allowed. This is used only by legacy event types. This option is not
+            available for new event types. Schema validation is lenient and allow fields that are not defined. Also,
+            it's possible to use the full json schema specification for defining schemas.
+
+          - 'none': Any schema modification is accepted, even if it might break existing producers or consumers.
+
+          - 'compatible': Schema changes never break existing consumers. Every event published since the first version
+            is still valid based on the newest schema. When in compatible mode, it's allowed to add new optional
+            properties and definitions to an existing schema, but no other changes are allowed. If additionalProperties
```

Comment:
@unlaut we could do that. Actually, I even developed this feature of blocking additionalProperties. But we have some users that rely on that for defining schemas that seems reasonable.

For example, let's suppose I have a schema that holds a Map<String, String>. Let's suppose it's the description of a product with its attributes.

```javascript
{
  "sku": "CM01-R",
  "product_attributes": {
    "color": "red",
    "size": "M"
  }
}
```

That could be achieved by specifying a schema like:

```javascript
{
  "properties": {
    "sku": {
      "type": "string",
    },
    "product_attributes": {
       "additionalProperties": {
         "type": "string"
       }
    }
  }
}
```

If we block additionalProperties altogether (which would make my life easier and also data lake more sane), user would be required to model such events in different ways. One alternative could be to use arrays, like:

```javascript
{
  "sku": "CM01-R",
  "product_attributes": [
    {
      "attribute_name": "color",      
      "attribute_value": "red"
    },
    {
      "attribute_name": "size",      
      "attribute_value": "M",      
    }
  }
}
```

Which schema would look like:

```javascript
{
  "properties": {
    "sku": {
      "type": "string",
    },
    "product_attributes": {
       "type": "array",
       "items": {
         "type": "object",
         "properties": {
           "attribute_name": {
             "type": "string"
           },
           "attribute_value": {
             "type": "string"
           }
         }
       }
    }
  }
}
```

That said, I think it would be completely possible to remove additionalProperties.